### DISCUSSION TITLE
Dynamic generate rules on the fly

### DISCUSSION THREAD

**Discussion by antaka1** [evrete/evrete#36] [Q&A] [answered]
Hello,

I have just discovered your repo and it looks amazing. I am trying to use it to build a set of rule engines which basically validates some steps for getting a product. For example I want to rent a car so there are few steps in this process:
- validate identity and age
- validate driving licence
- validate credit card
etc. 

This is just a silly example but it is just for context. Each of the steps above i can mimic them as a Rule and have something like this :
` Knowledge execute = service
                .newKnowledge()
                .newRule("Customer validation")
                .forEach("$i", Customer.class,
                        "$type", String.class,
                        "$o", ValidationResponse.class)
                .where("$type == 'IDENTITY_AGE'")
                .execute(ctx -> {
                    ....
                })
                .newRule("Driving validation")
                .forEach("$i", Driving.class,
                        "$type", String.class,
                        "$o", ValidationResponse.class)
                .where("$type == 'DRIVING_LICENCE'")
                .execute(ctx -> {
                    ....
}`

This looks nice but I want my app to support any type of products so i would like those rules to come from an API or from a database. 

One of the options i was thinking about is get a json / yaml with the product configuration and build a java class from it and then load the java class from an api. 

Is there any other way to build those rules dynamically ?

I went through all the documentation but still figured out if that's something supported out of the box.

Your opinion will be highly appreciated.

Thanks again for Evrete :)

Cristi

**Comment by andbi** [comment, marked as answer]
Hi! Basically, you are describing a DSL (Domain Specific Language), and the standard approach here is as follows:
1. You define the language for your rules and how they're stored (Excel, database, etc.)
2. When you need to convert a DSL rule source into Evrete's knowledge, you just parse the source and translate it into the very same `RuleBuilder` instructions you mentioned in your example.

The `RuleBuilder` interface is fluent, but that doesn't mean you can't "unchain" it and add conditions and actions only when your DSL source requires so. So you don't need to translate JSON/YAML into Java classes. Instead,  you create a new `Knowledge` instance, obtain a `RuleBuilder` from it, and apply conditions as you parse those JSON/YAML sources.

Regards,
@andbi

**Reply by antaka1** [reply]
Hey,

Thanks for coming back to me. I was looking over what you've told me and maybe my mind is not going really well now.

I am trying to do this:
`rules.forEach(r -> {
            knowledge
                    .newRule(r.getName())
                    .forEach(
                            "$i", VerificationResult.class,
                            "$steps", Steps.class
                    )
                    .where("r.getName().equals($i.getCheckType())")
                    .execute(ctx -> {
                        VerificationResult input = (VerificationResult) ctx.getObject("$i");
                        Steps s = (Steps) ctx.getObject("$steps");
                        Boolean evaluation = evaluateResults(input, r.getRequiredValidations());
                        if (evaluation.equals(Boolean.FALSE)) {
                            s.getSteps().add(Step.builder().name(r.getName()).build());
                        }
                    });
        });
        return knowledge;`

In the where I would like only one rule to be applied for the input, the one matching the checkType field ( eg Rule name = OTP_CHECK, checkType = OTP_CHECK ).

Is this possible in the where ? the other option is to remove where and add an if in the execute.

Thanks,
Cristi

**Comment by andbi** [comment]
Hi!
Conditions in the `execute` is definitely not the best way. Here is your reformatted code block:
```java
rules.forEach(r -> {
    knowledge
            .newRule(r.getName())
            .forEach("$i", VerificationResult.class, "$steps", Steps.class)
            .where("r.getName().equals($i.getCheckType())")
            .execute(ctx -> {
                VerificationResult input = ctx.get(VerificationResult.class, "$i");
                Steps s = ctx.get(Steps.class, "$steps");
                Boolean evaluation = evaluateResults(input, r.getRequiredValidations());
                if (evaluation.equals(Boolean.FALSE)) {
                    s.getSteps().add(Step.builder().name(r.getName()).build());
                }
            });
});
```
 
The `where` won't compile as the `r` variable won't be available to the compiler. So you will probably need to declare it in the `forEach` statement as well.

I'm also assuming that the `rules` var contains rules in your internal format. If so, then your code might look something like this:

```java
rules.forEach(r -> {
    knowledge
            .newRule(r.getName())
            .forEach("$i", VerificationResult.class, "$steps", Steps.class, "$rule", InternalRule.class)
            .where("$rule.getName().equals($i.getCheckType())")
            .execute(ctx -> {
                VerificationResult input = ctx.get(VerificationResult.class, "$i");
                Steps s = ctx.get(Steps.class, "$steps");
                InternalRule rule = ctx.get(InternalRule.class, "$rule");
                Boolean evaluation = evaluateResults(input, rule.getRequiredValidations());
                if (evaluation.equals(Boolean.FALSE)) {
                    s.getSteps().add(Step.builder().name(rule.getName()).build());
                }
            });
});
```
Just don't forget to  inject the internal rules  among other input values.
Is this what you are trying to achieve?

**Reply by antaka1** [reply]
Hey, thanks for the quick one.

So not really, the $rule == r. I cannot inject them as i am building the rules from a list. Maybe I am fully understanding the concepts here but i thought in forEach we are putting the objects we are iterating though, isn't it ?

So I have this
```
    Rule rule = Rule.builder()
            .name("CREDIT_SCORE")
            .order(2)
            .requiredValidations(RequiredValidations.builder()
                    .validationType(ValidationType.GTE_NUMBER_COMPARISON)
                    .expectedValue("80")
                    .build())
            .build();

    Rule rule2 = Rule.builder()
            .name("IDV")
            .order(1)
            .requiredValidations(RequiredValidations.builder()
                    .validationType(ValidationType.STRING_MATCH)
                    .expectedValue("PASSED")
                    .build())
            .build();
```

So i iterate though a list of this two rules and i want to build them. So this will be equivalent to:
```
 knowledge
            .newRule("IDV")
            .forEach("$i", VerificationResult.class, "$steps", Steps.class)
            .where("'IDV'.equals($i.getCheckType())")
            .execute(ctx -> {
             ......
            })
 .newRule("CREDIT_SCORE")
            .forEach("$i", VerificationResult.class, "$steps", Steps.class)
            .where("'CREDIT_SCORE'.equals($i.getCheckType())")
            .execute(ctx -> {
             ......
            });
```

Does it make more sense now what I am trying to achieve ?

Thanks,
Cristi

**Reply by andbi** [reply]
Now I am confused. You can not inject them but you use them in the `execute` blocks. What I proposed is a way of making these instances available to the engine's runtime context, that's it. 

PS: There is one extra condition I might've missed to guarantee that the `$rule == r` will effectively hold true:

```java
rules.forEach(r -> {
    knowledge
            .newRule(r.getName())
            .forEach("$i", VerificationResult.class, "$steps", Steps.class, "$rule", InternalRule.class)
            .where("$rule.getName().equals($i.getCheckType())")
            .where("$rule.getName().equals(\"" + r.getName() + "\")")
            .execute(ctx -> {
                // ...
            });
});
```

**Reply by antaka1** [reply]
Hello,

.where("$rule.getName().equals(\"" + r.getName() + "\")")

this was what I was missing. Thank you very much for your help :)
